worlds/lingo/player_logic.py

```diff
@@ -0,0 +1,233 @@
+                            self.handle_non_grouped_door(room_name, door_data, world, player, static_logic)
+
+        for room_name, room_data in StaticLingoLogic.DOORS_BY_ROOM.items():
+            for door_name, door_data in room_data.items():
+                if door_data.event:
+                    self.add_location(room_name, PlayerLocation(door_data.item_name, None, door_data.panels))
+                    self.EVENT_LOC_TO_ITEM[door_data.item_name] = door_data.item_name + " (Opened)"
+                    self.set_door_item(room_name, door_name, door_data.item_name + " (Opened)")
+
+        for room_name, room_data in StaticLingoLogic.PANELS_BY_ROOM.items():
+            for panel_name, panel_data in room_data.items():
+                if panel_data.achievement:
+                    event_name = room_name + " - " + panel_name + " (Achieved)"
+                    self.add_location(room_name, PlayerLocation(event_name, None,
+                                                                [RoomAndPanel(room_name, panel_name)]))
+                    self.EVENT_LOC_TO_ITEM[event_name] = "Mastery Achievement"
+
+        self.MASTERY_LOCATION = "Orange Tower Seventh Floor - THE MASTER"
+
+        if get_option_value(world, player, "victory_condition") == 0:
+            self.VICTORY_CONDITION = "Orange Tower Seventh Floor - THE END"
+            self.add_location("Orange Tower Seventh Floor", PlayerLocation("The End (Solved)"))
+            self.EVENT_LOC_TO_ITEM["The End (Solved)"] = "Victory"
+        elif get_option_value(world, player, "victory_condition") == 1:
+            self.VICTORY_CONDITION = "Orange Tower Seventh Floor - THE MASTER"
+            self.MASTERY_LOCATION = "Orange Tower Seventh Floor - Mastery Achievements"
+
+            self.add_location("Orange Tower Seventh Floor", PlayerLocation(self.MASTERY_LOCATION, None, []))
+            self.EVENT_LOC_TO_ITEM[self.MASTERY_LOCATION] = "Victory"
+
+        for location_name, location_data in StaticLingoLocations.ALL_LOCATION_TABLE.items():
+            if location_name != self.VICTORY_CONDITION:
+                if get_option_value(world, player, "reduce_checks")\
+                        and get_option_value(world, player, "shuffle_doors") == 0 and not location_data.include_reduce:
+                    continue
+
+                self.add_location(location_data.room, PlayerLocation(location_name, location_data.code,
+                                                                     location_data.panels))
+                self.REAL_LOCATIONS.append(location_name)
+
+        for name, item in StaticLingoItems.ALL_ITEM_TABLE.items():
+            if item.should_include(world, player):
+                self.REAL_ITEMS.append(name)
+
+        if get_option_value(world, player, "shuffle_paintings"):
+            # Shuffle paintings until we get something workable.
+            while not self.randomize_paintings(world, player, static_logic):
+                pass
```

Comment:
The chance of this requiring more than one call is very small (yet possible, which is why it's a loop). It would only happen if the generated painting mapping contained a cycle in required paintings, of which there are very few in the game (3 with doors, 7 with no doors). Plando wouldn't affect this because the painting mapping doesn't depend on any information from the multiworld (apart from the doors / no doors setting).

I can add a max counter and an exception, but I am also confident that this is safe.

Reply:
Actually, plando shouldn't factor into this.
Either way. It's just spooky ok :D

Reply:
I totally understand that.

Reply:
Addressed this, PTAL